src/LanguageServer/Impl/Documentation/DocstringConverter.cs

```diff
@@ -0,0 +1,460 @@
+            AppendLine(CurrentLineWithinBlock);
+            return true;
+        }
+
+        private bool ParseDoctest() {
+            if (CurrentLineIsOutsideBlock || string.IsNullOrWhiteSpace(CurrentLine)) {
+                TrimOutputAndAppendLine("```");
+                AppendLine();
+                PopState();
+                return false;
+            }
+
+            AppendLine(CurrentLineWithinBlock);
+            return true;
+        }
+
+        private bool BeginLiteralBlock() {
+            // The previous line must be empty.
+            var prev = LineAt(_lineNum - 1);
+            if (prev == null) {
+                return false;
+            } else if (!string.IsNullOrWhiteSpace(prev)) {
+                return false;
+            }
+
+            // Find the previous paragraph and check that it ends with ::
+            var i = _lineNum - 2;
+            for (; i >= 0; i--) {
+                var line = LineAt(i);
+
+                if (string.IsNullOrWhiteSpace(line)) {
+                    continue;
+                }
+
+                // Safe to ignore whitespace after the :: because all lines have been TrimEnd'd.
+                if (line.EndsWith("::")) {
+                    break;
+                }
+
+                return false;
+            }
+
+            if (i < 0) {
+                return false;
+            }
+
+            // Special case: allow one-liners at the same indent level.
+            if (CurrentIndent == 0) {
+                AppendLine("```");
+                PushAndSetState(ParseLiteralBlockSingleLine);
+                return true;
+            }
+
+            BeginMinIndentCodeBlock(ParseLiteralBlock);
+            return true;
+        }
+
+        private bool ParseLiteralBlock() {
+            // Slightly different than doctest, wait until the first non-empty unindented line to exit.
+            if (string.IsNullOrWhiteSpace(CurrentLine)) {
+                AppendLine();
+                return true;
+            }
+
+            if (CurrentLineIsOutsideBlock) {
+                TrimOutputAndAppendLine("```");
+                AppendLine();
+                PopState();
+                return false;
+            }
+
+            AppendLine(CurrentLineWithinBlock);
+            return true;
+        }
+
+        private bool ParseLiteralBlockSingleLine() {
+            AppendLine(CurrentLine);
+            AppendLine("```");
+            AppendLine();
+            PopState();
+            return true;
+        }
+
+        private bool BeginDirective() {
+            if (!Regex.IsMatch(CurrentLine, @"^\s*\.\. ")) {
+                return false;
+            }
+
+            PushAndSetState(ParseDirective);
+            _blockIndent = NextBlockIndent;
+            _appendDirectiveBlock = false;
+            return true;
+        }
+
+        private bool ParseDirective() {
+            // http://docutils.sourceforge.net/docs/ref/rst/restructuredtext.html#directives
```

Comment:
Do we need to add credits to license file?

Reply:
I'm not sure; I only used it as a reference (like we'd use the Python spec or similar). There's no actual parser described at that reference. These notes are just to point to the right places since the naming of certain terms in ReST is a bit hard to remember.